### PR TITLE
Adding the fokkerplanck_1d_4p3 (depends on 127)

### PR DESCRIPTION
**Pull request by dlg0**

### CODE REVIEW

File: src/pde.hpp [hunk 1]
```diff
@@ -33,7 +37,11 @@ enum class PDE_opts
   continuity_2,
   continuity_3,
   continuity_6,
+  fokkerplanck_1d_4p1a,
   fokkerplanck_1d_4p2,
+  fokkerplanck_1d_4p3,
```
Comment (bmcdanie) [inline src/pde.hpp]: that's a lot of fokkers

File: src/pde/pde_fokkerplanck1_4p3.hpp [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+  initial_condition_dim0(fk::vector<P> const x, P const t = 0)
+  {
+    ignore(t);
+    auto f = analytic_solution_dim0(x, 0);
+    return f;
+  }
+
+  // analytic solution
+
+  //  sig = 0.1;
```
Comment (bmcdanie) [inline src/pde/pde_fokkerplanck1_4p3.hpp]: do we want to leave this in? e.g. is it helpful in addition to the c++ code? for me, I'd rather just have one description of the pde, but I'm not dogmatic.

File: src/pde/pde_fokkerplanck1_4p4.hpp [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+      return std::exp(-std::pow(z - shift, 2) / std::pow(sig, 2));
+    case 3:
+      return std::exp(-std::pow(z + shift, 2) / std::pow(sig, 2));
+    case 4:
+      return std::exp(-std::pow(z - shift, 2) / std::pow(sig, 2)) +
+             std::exp(-std::pow(z + shift, 2) / std::pow(sig, 2));
+    }
+  }
+
+  //    function ret = soln(z,t)
```
Comment (bmcdanie) [inline src/pde/pde_fokkerplanck1_4p4.hpp]: same as above - do we want commented matlab in here?

File: src/pde/pde_fokkerplanck1_4p4.hpp [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+              false,                  // time-dependent
+              flux_type::downwind,    //
+              fk::vector<P>(),        // additional data vector
+              "d_dx",                 // name
+              dim0_                   // owning dim
+      );
+
+  inline static const std::vector<term<P>> termE = {termE_z};
+
+  // term 2
```
Comment (bmcdanie) [inline src/pde/pde_fokkerplanck1_4p4.hpp]: I'm starting to think maybe let's take it out

File: src/pde/pde_fokkerplanck1_4p5.hpp [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+              false,                  // time-dependent
+              flux_type::downwind,    //
+              fk::vector<P>(),        // additional data vector
+              "d_dx",                 // name
+              dim0_                   // owning dim
+      );
+
+  inline static const std::vector<term<P>> termE = {termE_z};
+
+  // term C
```
Comment (bmcdanie) [inline src/pde/pde_fokkerplanck1_4p5.hpp]: lots of this in here

File: src/pde.hpp (hunk 1 above)
Comment (dlg0) [reply src/pde.hpp]: LOL. Yeah, they could be made `RE_1d_*` if the fokkerplanck name is too long?

File: src/pde/pde_fokkerplanck1_4p5.hpp [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+
+  // g-funcs
+  static P g_func_0(P const x, P const time)
+  {
+    // suppress compiler warnings
+    ignore(x);
+    ignore(time);
+    return -1.0;
+  }
+  static P g_func_t1_z(P const x, P const time)
```
Comment (bmcdanie) [inline src/pde/pde_fokkerplanck1_4p5.hpp]: spell out term? I assume the z's have math meaning - are these matched to a dimension?

File: src/pde/pde_fokkerplanck1_4p5.hpp [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+              boundary_condition::neumann    // BCR_2
+      );
+
+  inline static const std::vector<term<P>> termC = {termC_z};
+
+  // term R
+  //
+  // - R d/dz(z(1-z^2) f)
+  //
+  // termR_z.type = 'grad'; % grad (see coeff_matrix.m for available types)
```
Comment (bmcdanie) [inline src/pde/pde_fokkerplanck1_4p5.hpp]: more matlab

File: src/pde.hpp (hunk 1 above)
Comment (bmcdanie) [reply src/pde.hpp]: I don't mind the length - in my professional physics opinion, fokkerplanck is a perfectly suitable name for these

File: src/pde/pde_fokkerplanck1_4p3.hpp [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+    }
+  }
+
+  static fk::vector<P>
+  analytic_solution_dim0(fk::vector<P> const z, P const t = 0)
+  {
+    fk::vector<P> f(z.size());
+    for (int i = 0; i < z.size(); ++i)
+    {
+      auto p  = phi(z(i), t);
```
Comment (bmcdanie) [inline src/pde/pde_fokkerplanck1_4p3.hpp]: const correctness - missed the first time...

File: src/pde/pde_fokkerplanck1_4p3.hpp [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+  // function definitions needed to build up the "dimension", "term", and
+  // "source" member objects below for this PDE
+  //
+
+  // specify initial condition vector functions...
+  static fk::vector<P>
+  initial_condition_dim0(fk::vector<P> const x, P const t = 0)
+  {
+    ignore(t);
+    auto f = analytic_solution_dim0(x, 0);
```
Comment (bmcdanie) [inline src/pde/pde_fokkerplanck1_4p3.hpp]: why store? just `return analytic_solution_dim0(x, 0);`

File: src/pde/pde_fokkerplanck1_4p3.hpp [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+  static P phi(P const z, P const t)
+  {
+    return z * std::exp(-t) /
+           std::sqrt(1 - (std::exp(-2 * t) - 1) * std::pow(z, 2));
+  }
+  static P f0(P const z)
+  {
+    static P const sig   = 0.1;
+    static P const shift = 0.36;
+    switch (4)
```
Comment (bmcdanie) [inline src/pde/pde_fokkerplanck1_4p3.hpp]: switching on a constant here?

File: src/pde/pde_fokkerplanck1_4p4.hpp [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+  // function definitions needed to build up the "dimension", "term", and
+  // "source" member objects below for this PDE
+  //
+
+  // specify initial condition vector functions...
+  static fk::vector<P>
+  initial_condition_dim0(fk::vector<P> const x, P const t = 0)
+  {
+    ignore(t);
+    auto f = analytic_solution_dim0(x, 0);
```
Comment (bmcdanie) [inline src/pde/pde_fokkerplanck1_4p4.hpp]: address these in all new files

File: src/pde/pde_fokkerplanck1_4p4.hpp [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+  }
+
+  static fk::vector<P>
+  analytic_solution_dim0(fk::vector<P> const z, P const t = 0)
+  {
+    ignore(t);
+    fk::vector<P> f(z.size());
+    for (int i = 0; i < z.size(); ++i)
+    {
+      auto A = E / C;
```
Comment (bmcdanie) [inline src/pde/pde_fokkerplanck1_4p4.hpp]: const

File: src/pde/pde_fokkerplanck1_4p3.hpp (hunk 8 above)
Comment (bmcdanie) [reply src/pde/pde_fokkerplanck1_4p3.hpp]: fixed

File: src/pde/pde_fokkerplanck1_4p3.hpp (hunk 10 above)
Comment (bmcdanie) [reply src/pde/pde_fokkerplanck1_4p3.hpp]: fixed.

File: src/pde/pde_fokkerplanck1_4p3.hpp (hunk 9 above)
Comment (bmcdanie) [reply src/pde/pde_fokkerplanck1_4p3.hpp]: fixed

File: src/pde/pde_fokkerplanck1_4p4.hpp (hunk 11 above)
Comment (bmcdanie) [reply src/pde/pde_fokkerplanck1_4p4.hpp]: fixed

File: src/pde/pde_fokkerplanck1_4p4.hpp (hunk 12 above)
Comment (bmcdanie) [reply src/pde/pde_fokkerplanck1_4p4.hpp]: fixed

**Pull request merged**